src/turbine/retransmit_service.zig

```diff
@@ -0,0 +1,430 @@
+/// 2. retransmitShreds: runs on N threads and receives RetransmitShredInfo's from the channel, computes the children to retransmit to
+///    and then constructs and sends packets to the network.
+pub fn run(
+    allocator: std.mem.Allocator,
+    my_contact_info: ThreadSafeContactInfo,
+    bank_fields: *const BankFields,
+    leader_schedule_cache: *LeaderScheduleCache,
+    gossip_table_rw: *RwMux(sig.gossip.GossipTable),
+    receiver: *Channel(Packet),
+    num_retransmit_sockets: usize,
+    maybe_num_retransmit_threads: ?usize,
+    overwrite_stake_for_testing: bool,
+    exit: *AtomicBool,
+    rand: Random,
+    logger: Logger,
+) !void {
+    errdefer {
+        logger.info().log("retransmit service failed");
+        exit.store(false, .monotonic);
+    }
+    const num_retransmit_threads = maybe_num_retransmit_threads orelse try std.Thread.getCpuCount();
+    logger.info().logf("starting retransmit service: num_retransmit_sockets={} num_retransmit_threads={}", .{
+        num_retransmit_sockets,
+        num_retransmit_threads,
+    });
+
+    var stats = try Stats.init();
+
+    var receive_to_retransmit_channel = try Channel(RetransmitShredInfo).init(allocator);
+    defer receive_to_retransmit_channel.deinit();
+
+    var retransmit_to_socket_channel = try Channel(Packet).init(allocator);
+    defer retransmit_to_socket_channel.deinit();
+
+    var retransmit_threads = std.ArrayList(std.Thread).init(allocator);
+    defer retransmit_threads.deinit();
+
+    var socket_threads = std.ArrayList(std.Thread).init(allocator);
+    defer socket_threads.deinit();
+
+    var retransmit_sockets: std.ArrayList(UdpSocket) = std.ArrayList(UdpSocket).init(allocator);
+    defer {
+        for (retransmit_sockets.items) |socket| socket.close();
+        retransmit_sockets.deinit();
+    }
+
+    for (0..num_retransmit_sockets) |_| {
+        var socket = try UdpSocket.create(.ipv4, .udp);
+        try socket.bind(try EndPoint.parse("0.0.0.0:0"));
+        try retransmit_sockets.append(try UdpSocket.create(.ipv4, .udp));
+    }
+
+    const receive_thread = try std.Thread.spawn(
+        .{},
+        receiveShreds,
+        .{
+            allocator,
+            my_contact_info,
+            bank_fields,
+            leader_schedule_cache,
+            receiver,
+            &receive_to_retransmit_channel,
+            gossip_table_rw,
+            rand,
+            exit,
+            logger,
+            &stats,
+            overwrite_stake_for_testing,
+        },
+    );
+
+    for (0..num_retransmit_threads) |_| {
+        try retransmit_threads.append(try std.Thread.spawn(
+            .{},
+            retransmitShreds,
+            .{
+                allocator,
+                &receive_to_retransmit_channel,
+                &retransmit_to_socket_channel,
+                &stats,
+                exit,
+            },
+        ));
+    }
+
+    for (retransmit_sockets.items) |socket| {
```

Comment:
are we sure we need multiple retransmit sockets?

Reply:
no not sure about this one, this was replicated from agave. Happy to remove since it has seemed to be fine running with a single socket

Reply:
Would you like me to remove them? Thumbs up is a little ambiguous in my mind

Reply:
e9478315